### PR TITLE
Feature/add auditing monkeypatch

### PR DESCRIPTION
**Pull request by Ethan-Arrowood**
## Description

This PR adds the preliminary implementation of extension based auditing. Looking forward to feedback as well as some input on how to test this. 

to do:
- [x] Add documentation
- [ ] Test through implementation (locally)
- [x] Test through unit tests

## Checklist for success

- [x] This change isn't already implemented by a pending pull request
- [x] This change is a child commit of a reasonably recent commit in the **master** branch 
    * Requests need not be a single commit, but should be a linear sequence of commits (i.e. no merge commits in your PR)
- [x] All tests pass locally (see [the testing section](https://github.com/azure/platform-chaos/blob/master/README.md#testing) for more info)
- [ ] All New behaviors or major functional changes have tests

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -83,6 +83,25 @@ Inflates the `resources` from a `req` objects `body` into a collection of object
 const objs = require('azure-chaos-fn/parsers').resourcesToObjects(req)
 ```
 
+### auditer
+
+A documented implementation of the verbose logging format defined in [Auditing](https://github.com/Azure/platform-chaos/wiki/Auditing)
+
+Initialize the auditer by using the following method:
```
Comment (bengreenier) [inline README.md]: we should talk more about what we're doing here - modifying context (and that it is azure specific) to help the user. Further we should clarify that there's some `index.auditDirectly` capability

File: audit.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+module.exports = function (context, opts) {
+  assert(opts, 'Options object must be defined')
+  assert(typeof opts.eventName === 'string', 'Event name must be a string')
+  assert(typeof opts.resource === 'string', 'Resource must be a string')
+
+  context['audit-options'] = opts
+  context['audit-queue'] = []
+  context['audit-system'] = `${os.hostname()}-${os.platform()}`
+
+  context['audit'] = extensionLogArgs => {
```
Comment (bengreenier) [inline audit.js]: we should avoid adding behavior to context, as it's something we don't own - consider a future update where azure decides context should have an audit property - we're now in the business of steamrolling it (bad ux, imho)

File: audit.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    const { eventName, resource } = context['audit-options']
+
+    context['audit-queue'].push({
```
Comment (bengreenier) [inline audit.js]: if we're going to keep this state on context, it should at least be escaped/unlikely to collide with user/platform features - like we did with `body['__audits']`

File: audit.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        body['__audits'] = audits
+        context.res.body = JSON.stringify(body)
+      } else if (typeof context.res.body === 'object') {
+        context.res.body['__audits'] = audits
+      } else {
+        /* Azure functions docs does not specify `res.body`
+         * can be anything besides a String or Object.
+         * If it can be another type it should be handeled here.
+         * */
+        throw new Error('Body is not string or object')
```
Comment (bengreenier) [inline audit.js]: we probably shouldn't throw here - this will cause a new side effect where `context.done()`  can fail

File: package.json [hunk 5]
```diff
@@ -20,6 +20,13 @@
     "eslint-plugin-node": "^7.0.1",
     "eslint-plugin-promise": "^4.0.0",
     "eslint-plugin-standard": "^4.0.0",
-    "mocha": "^5.2.0"
+    "lodash.isequal": "^4.5.0",
+    "mocha": "^5.2.0",
+    "sinon": "^6.2.0"
+  },
+  "dependencies": {
+    "pino": "^5.4.1",
```
Comment (bengreenier) [inline package.json]: remove

File: README.md [hunk 6]
```diff
@@ -1,7 +1,7 @@
 # platform-chaos
 
 
-[![Build Status](https://travis-ci.org/azure/platform-chaos.svg?branch=master)](https://travis-ci.org/azure/platform-chaos)
+[![Build Status](https://travis-ci.org/Azure/platform-chaos.svg?branch=master)](https://travis-ci.org/Azure/platform-chaos)
```
Comment (allydurks) [inline README.md]: :sparkles:

File: audit.js [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+      if (typeof context.res.body === 'string') {
+        const body = JSON.parse(context.res.body)
+        body['__audits'] = audits
+        context.res.body = JSON.stringify(body)
+      } else if (typeof context.res.body === 'object') {
+        context.res.body['__audits'] = audits
+      } else {
+        /* Azure functions docs does not specify `res.body`
+         * can be anything besides a String or Object.
+         * If it can be another type it should be handeled here.
```
Comment (allydurks) [inline audit.js]: [nit] handled

File: audit.js [hunk 8]
```diff
@@ -3,54 +3,61 @@ const assert = require('assert')
...
+    auditId: uuidv4(),
+    eventName: eventName,
+    system: auditSystem,
+    date: new Date().toISOString(),
+    resource: resource,
+    extensionLog: typeof extensionLogArgs === 'string' ? [ extensionLogArgs ] : Array.from(extensionLogArgs)
   })
+}
+
+module.exports = {
```
Comment (bengreenier) [inline audit.js]: i don't think we should actually call this `auditDirectly`, that was an example sorry for any confusion! maybe `audit` or similar depending on the verbiage expected when a dev references this module

File: audit.js [hunk 9]
```diff
@@ -3,54 +3,61 @@ const assert = require('assert')
...
-        context.res.body = JSON.stringify(body)
-      } else if (typeof context.res.body === 'object') {
-        context.res.body['__audits'] = audits
-      } else {
-        /* Azure functions docs does not specify `res.body`
-         * can be anything besides a String or Object.
-         * If it can be another type it should be handeled here.
-         * */
-        throw new Error('Body is not string or object')
+    // clearQueue()
```
Comment (bengreenier) [inline audit.js]: if this shouldn't be here please remove

**Comment by Ethan-Arrowood** [comment]
My last commit should clean up some of Ben's code changes. I'm running into some issues developing a functional example, but still working at it. I'm going to hold off on doc changes until I figure out implementation.

Also what should I do instead of throwing an error? Could I use console.warn or something like that?

**Comment by Ethan-Arrowood** [comment]
🎉woo! implementation works! I modified the code from the tutorial wiki with my auditer (used `npm link` to use local version of package). Notice in the top right there is a list of 'audits' being output from line 30 of the file on the left. Thomas' published chaos event is sorta broken, tomorrow I'll be able to focus on a pr updating it.

Only thing left for this PR is documentation and deciding what to do about the error line.

auditdemo

**Comment by Ethan-Arrowood** [comment]
I have opened a PR on Thomas' extension that once merged will reference in the documentation for this feature addition https://github.com/trstringer/azure-chaos-fn-webapp-startstop/pull/1

**Comment by Ethan-Arrowood** [comment]
Getting close on this one. I'd like to include an example extension (Thomas' thing) in the documentation. Now that I've settled the --resources issue I have to retest everything and make sure it is working as expected too.

File: README.md [hunk 10]
```diff
@@ -85,20 +85,18 @@ const objs = require('azure-chaos-fn/parsers').resourcesToObjects(req)
...
 
 Initialize the auditer by using the following method:
 ```js
 const index = require('azure-chaos-fn')
 
 index.auditer(/* Azure Function context */, {
-    eventName: /* Chaos event name */,
-    resource: /* Target resource */
+    eventName: /* Chaos event name : string */,
+    resources: /* Target resources : string */
```
Comment (bengreenier) [inline README.md]: string or array?

**Pull request merged**